mlir/python/PyCompilerDriver.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "Catalyst/Driver/CompilerDriver.h"
+#include "mlir/Bindings/Python/PybindAdaptors.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "llvm/Support/raw_ostream.h"
+#include <iostream>
+
+namespace py = pybind11;
+using namespace mlir::python::adaptors;
+
+std::vector<Pipeline> parseCompilerSpec(const py::list &pipelines)
+{
+    std::vector<Pipeline> out;
+    for (py::handle obj : pipelines) {
+        py::tuple t = obj.cast<py::tuple>();
+        auto i = t.begin();
+        auto py_name = i;
+        i++;
+        auto py_passes = i;
+        i++;
+        assert(i == t.end());
+        std::string name = py_name->attr("__str__")().cast<std::string>();
+        Pipeline::PassList passes;
+        std::transform(py_passes->begin(), py_passes->end(), std::back_inserter(passes),
+                       [](py::handle p) { return p.attr("__str__")().cast<std::string>(); });
+        out.push_back(Pipeline({name, passes}));
+    }
+    return out;
+}
+
+PYBIND11_MODULE(_catalystDriver, m)
+{
+    //===--------------------------------------------------------------------===//
+    // Catalyst Compiler Driver
+    //===--------------------------------------------------------------------===//
+    py::class_<FunctionAttributes> funcattrs_class(m, "FunctionAttributes");
+    funcattrs_class.def(py::init<>())
+        .def("getFunctionName",
+             [](const FunctionAttributes &fa) -> std::string { return fa.functionName; })
+        .def("getReturnType",
+             [](const FunctionAttributes &fa) -> std::string { return fa.returnType; });
+
+    py::class_<CompilerOutput> compout_class(m, "CompilerOutput");
+    compout_class.def(py::init<>())
+        .def("get_pipeline_output",
+             [](const CompilerOutput &co, const std::string &name) -> std::optional<std::string> {
+                 auto res = co.pipelineOutputs.find(name);
+                 return res != co.pipelineOutputs.end() ? res->second
+                                                        : std::optional<std::string>();
+             })
+        .def("get_output_IR", [](const CompilerOutput &co) -> std::string { return co.outIR; })
```

Comment:
nit: I'd recommend using lowercase snake_case names because it's a Python interface (and for consistency with the other methods below).
```suggestion
        .def("get_function_name",
             [](const FunctionAttributes &fa) -> std::string { return fa.functionName; })
        .def("get_return_type",
             [](const FunctionAttributes &fa) -> std::string { return fa.returnType; });

    py::class_<CompilerOutput> compout_class(m, "CompilerOutput");
    compout_class.def(py::init<>())
        .def("get_pipeline_output",
             [](const CompilerOutput &co, const std::string &name) -> std::optional<std::string> {
                 auto res = co.pipelineOutputs.find(name);
                 return res != co.pipelineOutputs.end() ? res->second
                                                        : std::optional<std::string>();
             })
        .def("get_output_ir", [](const CompilerOutput &co) -> std::string { return co.outIR; })
```